sqlglot/lineage.py

```diff
@@ -276,8 +276,44 @@ def to_node(
                 reference_node_name=reference_node_name,
                 trim_selects=trim_selects,
             )
+
+        elif scope.pivots:
+            pivot = next((p for p in scope.pivots if p.alias_or_name == c.table), None)
+            if pivot:
+                # The source is a pivot operation, so we need to trace back to the aggregated column
+                pivot_column_mapping = {}
+                for agg in pivot.expressions:
+                    agg_columns = list(agg.find_all(exp.Column))
+                    literals = list(pivot.args.get("field").find_all(exp.Literal))
+                    for literal in literals:
+                        pivot_column_name = f"{literal.this}"
+                        if isinstance(agg, exp.Alias):
+                            pivot_column_name += f"_{agg.alias_or_name}"
+                        pivot_column_mapping[pivot_column_name] = agg_columns
```

Comment:
Is this necessary? Note that the parser takes care of inferring the pivot column names for each dialect, take a look here, it may be possible to simplify this section.

https://github.com/tobymao/sqlglot/blob/3945acc4a0dfd58147de929c9a2c71734d8f1ade/tests/test_parser.py#L596-L712

Reply:
I know what you mean.
I could not find a more simplified way, because the `Pivot` expression has the `columns` in `args`, but these are only `Identifier`. And I do not know how to get from these identifiers to the expressions which contain the aggregates.

Maybe you can give me a hint to simplify this properly?

Reply:
Does it make sense to extend the parsing to keep that information together somewhere here:
https://github.com/tobymao/sqlglot/blob/a9dca8dd1b523efd703003694d4389f9af9d1a12/sqlglot/parser.py#L4090C1-L4104C42

Reply:
Perhaps it does, @VaggelisD is also taking a look at this so we should have more constructive feedback soon.

Reply:
If I understand correctly, this could be improved if we take advantage of how `pivot.columns` are generated, e.g:

```
>>> sql = """
SELECT prop1_p as other_a
FROM (SELECT partname, price, quality FROM part)
PIVOT (
    AVG(price) AS p, MAX(quality) AS q FOR partname IN ('prop' AS prop1, 'rudder', 'test')
);
"""

>> ast = sqlglot.parse_one(sql, dialect="duckdb")
>> print(repr(ast))
Select(
  expressions=[...],
  from=From(...),
      pivots=[
        Pivot(
          ...,
          columns=[
            Identifier(this=prop1_p, quoted=False),
            Identifier(this=prop1_q, quoted=False),
            Identifier(this=rudder_p, quoted=False),
            Identifier(this=rudder_q, quoted=False),
            Identifier(this=test_p, quoted=False),
            Identifier(this=test_q, quoted=False)])])))
```

Note how the order is alternating between each aggfunc & field; That means that for each aggfunc, it'll prefix/suffix the the field name in a "step" manner. For instance, `AVG(price) AS p` suffixes the fields in indices `(0, 2, 4, ...)` and so on. The first for-loop could then be generalized like:

```Python3
        elif scope.pivots:
            pivot = next(
                (p for p in scope.pivots if p.alias_or_name == c.table and not p.unpivot), None
            )
            if pivot:
                # The source is a pivot operation, so we need to trace back to the aggregated column
                pivot_column_mapping = {}
                pivot_aggs = len(pivot.expressions)
                columns = pivot.args.get("columns")

                for i, agg in enumerate(pivot.expressions):
                    agg_cols = list(agg.find_all(exp.Column))
                    for col in range(i, len(columns), pivot_aggs):
                        # Go through the field_names in a step manner for this aggfunc
                        pivot_column_mapping[columns[col].name] = agg_cols

                for agg_column in pivot_column_mapping[c.alias_or_name]:
                    ...
```

With that change, all of the tests (including the new ones) pass, and it can even be further simplified.

Reply:
Yes that simplifies it quite a bit.
I added this change to the PR.